src/driver-asus.c

```diff
@@ -0,0 +1,516 @@
+		continue;
+
+	log_debug(device->ratbag, "Profile %d changed\n", profile->index);
+
+	// switch profile
+	if (profile->index != profile_id) {
+		log_debug(device->ratbag, "Switching to profile %d\n", profile->index);
+		error = asus_set_profile(device, profile->index);
+		if (error)
+			return error;
+	}
+
+	// set buttons
+	ratbag_profile_for_each_button(profile, button) {
+		if (!button->dirty)
+			continue;
+
+		asus_index = mapping[button->index] != -1 ? mapping[button->index] : ASUS_BUTTON_MAPPING_ORDER[button->index];
+		if (asus_index == -1) {
+			log_debug(device->ratbag, "No mapping for button %d\n", button->index);
+			continue;
+		}
+
+		log_debug(device->ratbag, "Button %d (%d) changed\n", button->index, asus_index);
+		error = 0;
+
+		switch(button->action.type) {
+		case RATBAG_BUTTON_ACTION_TYPE_NONE:
+			error = asus_set_button_action(
+				device, asus_index,
+				ASUS_BUTTON_CODE_DISABLED,
+				ASUS_BUTTON_ACTION_TYPE_BUTTON);
+			break;
+
+		case RATBAG_BUTTON_ACTION_TYPE_KEY:
+			// linux code to asus code
+			asus_index_to = asus_find_key_index_by_code(button->action.action.key.key);
+			if (asus_index_to >= 0)
+				error = asus_set_button_action(
+					device, asus_index,
+					asus_index_to,
+					ASUS_BUTTON_ACTION_TYPE_KEY);
+		break;
+
+		case RATBAG_BUTTON_ACTION_TYPE_BUTTON:
+		case RATBAG_BUTTON_ACTION_TYPE_SPECIAL:
+			asus_index_to = asus_find_button_index_by_action(button->action);
+			if (asus_index_to >= 0)
+				error = asus_set_button_action(
+					device, asus_index,
+					ASUS_BUTTON_MAPPING[asus_index_to].asus_code,
+					ASUS_BUTTON_ACTION_TYPE_BUTTON);
+		break;
+
+		default:
+			break;
+		}
+
+		if (error)
+			return error;
+	}
+
+	// set extra settings
+	if (profile->rate_dirty) {
+		log_debug(device->ratbag, "Polling rate changed to %d Hz\n", profile->hz);
+		error = asus_set_polling_rate(device, profile->hz);
+		if (error)
+			return error;
+	}
+
+	// set DPIs
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+
+		log_debug(
+			device->ratbag, "Resolution %d changed to %d\n",
+			resolution->index, resolution->dpi_x);
+
+		error = asus_set_dpi(device, resolution->index, resolution->dpi_x);
+		if (error)
+			return error;
+	}
+
+	// set LEDs
+	if (led_count) {
+		ratbag_profile_for_each_led(profile, led) {
+			if (!led->dirty)
+				continue;
+
+			log_debug(device->ratbag, "LED %d changed\n", led->index);
+			led_mode = 0;
+			for (i = 0; i < ARRAY_LENGTH(ASUS_LED_MODE); i++) {
+				if (ASUS_LED_MODE[i] == led->mode) {
+					led_mode = i;
+					break;
+				}
+			}
+
+			// convert brightness from 0-256 to 0-4
+			brightness = (uint8_t) round((double) led->brightness / 64.0);
+			error = asus_set_led(device, led->index, led_mode, brightness, led->color);
+			if (error)
+				return error;
+		}
+	}
+
+		// save profile
+		log_debug(device->ratbag, "Saving profile\n");
+		error = asus_save_profile(device);
+		if (error)
+			return error;
+	}
```

Comment:
something's wrong with the indentation for this block to be indented - line 270, the profile loop.

Might be easiest to split this function into the one that loops of the profile and one helper function that takes the profile as argument and does the work just for this profile (i.e. the loop body as one function). This way it ends up being:
```c
int asus_driver_save_profiles(...) {
    ratbag_device_for_each_profile(device, profile) {
	if (!profile->dirty)
		continue;
        asus_driver_save_profile(profile);
   }
}
```